File: backend/controllers/question_bank_controller.py
from flask import request, jsonify, g, send_file
from controllers import handle_errors
import services.question_bank_service as svc


@handle_errors
def list_questions():
    if g.user["role"] != "teacher":
        return jsonify({"success": False, "message": "Không có quyền"}), 403
    q          = request.args.get("q", "").strip()
    page       = request.args.get("page",  1,  type=int)
    limit      = request.args.get("limit", 10, type=int)
    lesson_raw = request.args.get("lesson_id")
    lesson_id  = int(lesson_raw) if lesson_raw is not None else None
    return jsonify({"success": True, "data": svc.list_questions(g.user["user_id"], q, page, limit, lesson_id=lesson_id)})


@handle_errors
def create_question():
    if g.user["role"] != "teacher":
        return jsonify({"success": False, "message": "Không có quyền"}), 403
    data    = request.get_json() or {}
    content = data.get("content", "").strip()
    if not content:
        return jsonify({"success": False, "message": "Nội dung câu hỏi không được trống"}), 400
    lesson_id = data.get("lesson_id") or None
    new_id = svc.create_question(
        g.user["user_id"], content,
        data.get("explanation") or None,
        data.get("answers", []),
        lesson_id=lesson_id,
    )
    return jsonify({"success": True, "id": new_id}), 201


@handle_errors
def update_question(question_id):
    if g.user["role"] != "teacher":
        return jsonify({"success": False, "message": "Không có quyền"}), 403
    data    = request.get_json() or {}
    content = data.get("content", "").strip()
    if not content:
        return jsonify({"success": False, "message": "Nội dung câu hỏi không được trống"}), 400
    lesson_id = data.get("lesson_id") or None
    svc.update_question(
        question_id, g.user["user_id"], content,
        data.get("explanation") or None,
        data.get("answers", []),
        lesson_id=lesson_id,
    )
    return jsonify({"success": True, "message": "Đã cập nhật câu hỏi"})


@handle_errors
def download_sample():
    buf = svc.generate_sample_excel()
    return send_file(
        buf,
        mimetype="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        as_attachment=True,
        download_name="mau_ngan_hang_cau_hoi.xlsx",
    )


@handle_errors
def import_questions():
    if g.user["role"] != "teacher":
        return jsonify({"success": False, "message": "Không có quyền"}), 403
    if "file" not in request.files:
        return jsonify({"success": False, "message": "Không tìm thấy file"}), 400
    file = request.files["file"]
    if not file.filename or not file.filename.lower().endswith((".xlsx", ".xls")):
        return jsonify({"success": False, "message": "Chỉ hỗ trợ file .xlsx hoặc .xls"}), 400
    lesson_raw = request.form.get("lesson_id")
    lesson_id  = int(lesson_raw) if lesson_raw else None
    result = svc.import_questions_from_excel(g.user["user_id"], file.read(), lesson_id=lesson_id)
    return jsonify({"success": True, "data": result})


@handle_errors
def delete_question(question_id):
    if g.user["role"] != "teacher":
        return jsonify({"success": False, "message": "Không có quyền"}), 403
    svc.delete_question(question_id, g.user["user_id"])
    return jsonify({"success": True, "message": "Đã xóa câu hỏi"})

File: backend/src/controllers/question-bank.controller.js
import * as svc from '../services/question-bank.service.js';
import { success, created, successMsg } from '../utils/response.js';

export const listQuestions = async (req, res) => {
  const q = req.query.q || '';
  const page = parseInt(req.query.page) || 1;
  const limit = parseInt(req.query.limit) || 10;
  const lessonRaw = req.query.lessonId;
  const lessonId = lessonRaw !== undefined ? parseInt(lessonRaw) : undefined;
  success(res, await svc.listQuestions(req.user.id, q, page, limit, lessonId));
};

export const createQuestion = async (req, res) => {
  const { content, explanation = null, answers = [], lessonId } = req.body;
  const id = await svc.createQuestion(req.user.id, content.trim(), explanation || null, answers, lessonId);
  created(res, { id });
};

export const updateQuestion = async (req, res) => {
  const { content, explanation = null, answers = [], lessonId } = req.body;
  await svc.updateQuestion(parseInt(req.params.id), req.user.id, content.trim(), explanation || null, answers, lessonId);
  successMsg(res, 'Đã cập nhật câu hỏi');
};

export const deleteQuestion = async (req, res) => {
  await svc.deleteQuestion(parseInt(req.params.id), req.user.id);
  successMsg(res, 'Đã xóa câu hỏi');
};

export const generateQuestions = async (req, res) => {
  const { lessonId, numQuestions, description } = req.body;
  const result = await svc.generateQuestions(req.user.id, lessonId, numQuestions, description);
  success(res, result);
};

export const saveQuestionsBatch = async (req, res) => {
  const { lessonId, questions } = req.body;
  const saved = await svc.saveBatch(req.user.id, questions, lessonId);
  created(res, { saved }, `Đã lưu ${saved} câu hỏi vào ngân hàng`);
};

export const deleteQuestions = async (req, res) => {
  const { deleted, requested } = await svc.deleteQuestions(req.body.ids, req.user.id);
  success(res, { deleted, requested }, `Đã xóa ${deleted} câu hỏi`);
};

export const downloadSample = async (req, res) => {
  const buffer = svc.generateSampleExcel();
  res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
  res.setHeader('Content-Disposition', 'attachment; filename="mau_ngan_hang_cau_hoi.xlsx"');
  res.send(buffer);
};

export const importQuestions = async (req, res) => {
  if (!req.file) return res.status(400).json({ success: false, message: 'Không tìm thấy file' });
  // Import đi qua multipart nên không dùng được Zod như 2 route kia — chặn tay ở đây.
  const lessonId = parseInt(req.body?.lessonId);
  if (!Number.isInteger(lessonId) || lessonId <= 0) {
    return res.status(400).json({ success: false, message: 'Vui lòng chọn chủ đề trước khi import' });
  }
  const result = await svc.importQuestionsFromExcel(req.user.id, req.file.buffer, lessonId);
  success(res, result);
};
